src/integrationTest/sam.test.ts

```diff
@@ -186,6 +186,86 @@ async function getAddConfigCodeLens(documentUri: vscode.Uri): Promise<vscode.Cod
     }
 }
 
+/**
+ * Returns a string if there is a validation issue, undefined if there is no issue.
+ */
+function validateSamDebugSession(
+    debugSession: vscode.DebugSession,
+    expectedName: string,
+    expectedRuntime: string
+): string | undefined {
+    const runtime = (debugSession.configuration as any).runtime
+    const name = (debugSession.configuration as any).name
+    if (name !== expectedName || runtime !== expectedRuntime) {
+        const failMsg =
+            `Unexpected DebugSession (expected name="${expectedName}" runtime="${expectedRuntime}"):` +
+            `\n${JSON.stringify(debugSession)}`
+        return failMsg
+    }
+}
+
+/**
+ * Simulates pressing 'F5' to start debugging. Sets up events to see if debugging was successful
+ * or not. Since we are not checking outputs we treat a successful operation as the debug session
+ * closing on its own (i.e. the container executable terminated)
+ *
+ * @param scenario Scenario to run, used for logging information
+ * @param scenarioIndex Scenario number, used for logging information
+ * @param testConfig Debug configuration to start the debugging with
+ * @param testDisposables All events registered by this function are pushed here to be removed later
+ * @param sessionLog An array where session logs are stored
+ */
+async function startDebugger(
+    scenario: TestScenario,
+    scenarioIndex: number,
+    testConfig: vscode.DebugConfiguration,
+    testDisposables: vscode.Disposable[],
+    sessionLog: string[]
+) {
+    // Create a Promise that encapsulates our success critera
+    const success = new Promise<void>((resolve, reject) => {
+        testDisposables.push(
+            vscode.debug.onDidTerminateDebugSession(async endedSession => {
+                sessionLog.push(`scenario ${scenarioIndex} (END) (runtime=${scenario.runtime}) ${endedSession.name}`)
+                const sessionRuntime = (endedSession.configuration as any).runtime
+                if (!sessionRuntime) {
+                    // It's a coprocess, ignore it.
+                    return
+                }
+                const failMsg = validateSamDebugSession(endedSession, testConfig.name, scenario.runtime)
+                if (failMsg) {
+                    reject(new Error(failMsg))
+                }
+                resolve()
+                await stopDebugger(`${scenario.runtime} / onDidTerminateDebugSession`)
+            })
+        )
+    })
+
+    // Executes the 'F5' action
+    await vscode.debug.startDebugging(undefined, testConfig).then(
```

Comment:
Is `await...then` redundant?

Reply:
No, then returns a Promise. Since we aren't explicitly returning anything in this code it is needed (you could change the `await` to a `return` and it would work fine). It would also be needed if you wanted to run clean-up code after `startDebugging`. Promises are kind of tricky since a lot of things are inferred by the transpiler when you use them.

Reply:
Am I missing something? Via [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function):
> Code after each await expression can be thought of as existing in a .then callback. In this way a promise chain is progressively constructed with each reentrant step through the function. The return value forms the final link in the chain.